crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -1003,6 +1017,260 @@ pub fn verify_nsec(query: &Query, soa_name: &Name, nsecs: &[&Record]) -> Proof {
+///    CNAME set in its Type Bit Maps field.
+///
+///    If there is no such NSEC3 RR, then the validator MUST verify that a
+///    closest provable encloser proof for QNAME is present in the response,
+///    and that the NSEC3 RR that covers the "next closer" name has the Opt-
+///    Out bit set.
+///
+/// 8.7.  Validating Wildcard No Data Responses
+///
+///    The validator MUST verify a closest encloser proof for QNAME and MUST
+///    find an NSEC3 RR present in the response that matches the wildcard
+///    name generated by prepending the asterisk label to the closest
+///    encloser.  Furthermore, the bits corresponding to both QTYPE and
+///    CNAME MUST NOT be set in the wildcard matching NSEC3 RR.
+///
+/// 8.8.  Validating Wildcard Answer Responses
+///
+///    The verified wildcard answer RRSet in the response provides the
+///    validator with a (candidate) closest encloser for QNAME.  This
+///    closest encloser is the immediate ancestor to the generating
+///    wildcard.
+///
+///    Validators MUST verify that there is an NSEC3 RR that covers the
+///    "next closer" name to QNAME present in the response.  This proves
+///    that QNAME itself did not exist and that the correct wildcard was
+///    used to generate the response.
+///
+/// 8.9.  Validating Referrals to Unsigned Subzones
+///
+///    The delegation name in a referral is the owner name of the NS RRSet
+///    present in the authority section of the referral response.
+///
+///    If there is an NSEC3 RR present in the response that matches the
+///    delegation name, then the validator MUST ensure that the NS bit is
+///    set and that the DS bit is not set in the Type Bit Maps field of the
+///    NSEC3 RR.  The validator MUST also ensure that the NSEC3 RR is from
+///    the correct (i.e., parent) zone.  This is done by ensuring that the
+///    SOA bit is not set in the Type Bit Maps field of this NSEC3 RR.
+///
+///    Note that the presence of an NS bit implies the absence of a DNAME
+///    bit, so there is no need to check for the DNAME bit in the Type Bit
+///    Maps field of the NSEC3 RR.
+///
+///    If there is no NSEC3 RR present that matches the delegation name,
+///    then the validator MUST verify a closest provable encloser proof for
+///    the delegation name.  The validator MUST verify that the Opt-Out bit
+///    is set in the NSEC3 RR that covers the "next closer" name to the
+///    delegation name.
+///
+#[doc(hidden)]
+pub fn verify_nsec3(query: &Query, _soa_name: &Name, nsec3s: &[&Record]) -> Proof {
+    fn hash(name: &Name, salt: &[u8], iterations: u16) -> Vec<u8> {
+        Nsec3HashAlgorithm::SHA1
+            .hash(salt, name, iterations)
+            .expect("SHA1 Hashing of a domanin name failed")
+            .to_vec()
+    }
+
+    debug_assert!(!nsec3s.is_empty());
+    debug_assert!(nsec3s.iter().all(|rr| is_dnssec(rr, RecordType::NSEC3)));
+
+    let nsec3s: Vec<_> = nsec3s
```

Comment:
Is it possible that this `nsec3s` can be empty afterwards? If yes then the below access of `nsec3s[0]` will panic.

Reply:
This is prevented by debug asserts, and from the calcite you can see that we only call verify_nsec3 function in a situation when there are some NSEC3 records present.

`debug_assert!`s above are there more as a failsafe in case the top-level validation function gets changed in future and the invariant would be broken. Any client NSEC3 test would pass through these asserts so we will know about the invariant being broken when the hypothetical new change hits CI.